### PR TITLE
feat: (PRO-413) Enhance fee payer outflow calculation to include SPL …

### PR DESCRIPTION
**Pull request by dev-jodee**
…token transfers

- Updated `calculate_fee_payer_outflow` to account for SPL token transfers, improving accuracy in fee estimation.
- Added comprehensive tests for SPL token transfer scenarios to ensure correct fee calculations.
- Adjusted transaction validation to utilize updated outflow calculations, ensuring consistency across transaction processing.

----

> [!IMPORTANT]
> Enhance fee payer outflow calculation to include SPL token transfers for accurate fee estimation and update related transaction validation and tests.
> 
>   - **Behavior**:
>     - Update `calculate_fee_payer_outflow` in `fee.rs` to include SPL token transfers for accurate fee estimation.
>     - Adjust transaction validation in `transaction_validator.rs` to use updated outflow calculations.
>   - **Testing**:
>     - Add tests in `fee.rs` for SPL token transfer scenarios to ensure correct fee calculations.
>     - Add integration tests in `jupiter_integration.rs` for unknown tokens.
>   - **Misc**:
>     - Minor error handling improvements in `oracle.rs` and `auth.rs`.
>     - Update `encode_versioned_transaction` in `transaction.rs` to return `Result`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=solana-foundation%2Fkora&utm_source=github&utm_medium=referral)<sup> for 22d4f9218a3a2f7026e3b8fa0c4d1d27ef313715. You can [customize](https://app.ellipsis.dev/solana-foundation/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

----

## 📊 Unit Test Coverage
Coverage

**Unit Test Coverage: 83.7%**

[View Detailed Coverage Report](https://github.com/solana-foundation/kora/actions/runs/18541937461)

### CODE REVIEW

File: crates/lib/src/token/token.rs [hunk 1]
```diff
@@ -140,6 +140,119 @@ impl TokenUtil {
...
+                    // Check if fee payer owns the destination (inflow)
+                    // We need to check the destination token account owner
+                    if let Some(mint_pubkey) = mint {
+                        // Get destination account to check owner
+                        match CacheUtil::get_account(rpc_client, destination_address, false).await {
+                            Ok(dest_account) => {
+                                let token_program =
+                                    TokenType::get_token_program_from_owner(&dest_account.owner)?;
+                                if let Ok(token_account) =
+                                    token_program.unpack_token_account(&dest_account.data)
```
Comment (amilz) [inline crates/lib/src/token/token.rs:182]: Random thought, but i think T22 is backwards compatible here, so we could just always use T22 to unpack (we'd want to test to be sure, but pretty sure it just 0's out the extra data).

File: crates/lib/src/token/token.rs [hunk 2]
```diff
@@ -140,6 +140,119 @@ impl TokenUtil {
...
+                                    if token_account.owner() == *fee_payer {
+                                        mint_to_transfers
+                                            .entry(*mint_pubkey)
+                                            .or_default()
+                                            .push((*amount, false)); // inflow
+                                    }
+                                }
+                            }
+                            Err(_) => {
+                                // Skip if destination account doesn't exist or can't be fetched
```
Comment (amilz) [inline crates/lib/src/token/token.rs]: Is this okay? Suppose we have an ATA creation ix prior to this one--won't we skip b/c it's not found? maybe an alt to the `get_account` is to just derive atas and make sure it's valid ata

File: crates/lib/src/token/token.rs [hunk 3]
```diff
@@ -140,6 +140,119 @@ impl TokenUtil {
...
+            3,
+            Duration::from_secs(1),
+            get_price_oracle(price_source.clone()),
+        );
+
+        let prices = oracle.get_token_prices(&mint_addresses).await?;
+
+        let mut mint_decimals = std::collections::HashMap::new();
+        for mint in mint_to_transfers.keys() {
+            let decimals = Self::get_mint_decimals(rpc_client, mint).await?;
```
Comment (amilz) [inline crates/lib/src/token/token.rs:249]: this is maybe a longer term todo, but i wonder if we should have varying ttl for mint if we're just using to get owner/decimals...shouldn't ever really have to fetch fresh? wdyt

File: crates/lib/src/oracle/jupiter.rs [hunk 4]
```diff
@@ -81,13 +81,29 @@ impl PriceOracle for JupiterPriceOracle {
...
+    ) -> Result<HashMap<String, TokenPrice>, KoraError> {
+        if mint_addresses.is_empty() {
+            return Ok(HashMap::new());
+        }
+
         // Try pro API first if API key is available, then fallback to free API
         if let Some(api_key) = &self.api_key {
             match self
-                .fetch_price_from_url(client, &self.pro_api_url, mint_address, Some(api_key))
+                .fetch_prices_from_url(client, &self.pro_api_url, mint_addresses, Some(api_key))
```
Comment (amilz) [inline crates/lib/src/oracle/jupiter.rs:104]: one thing we need to test is what happens if an unknown mint gets passed to jup? does it throw or just included some null data for that mint?

File: crates/lib/src/oracle/jupiter.rs [hunk 5]
```diff
@@ -143,16 +166,33 @@ impl JupiterPriceOracle {
...
-        let price_data = jupiter_response
-            .get(mint_address)
-            .ok_or_else(|| KoraError::RpcError("No price data from Jupiter".to_string()))?;
 
-        let price = price_data.usd_price / sol_price.usd_price;
+        // Convert all prices to SOL-denominated
+        let mut result = HashMap::new();
+        for mint_address in mint_addresses {
+            if let Some(price_data) = jupiter_response.get(mint_address.as_str()) {
+                let price = price_data.usd_price / sol_price.usd_price;
```
Comment (amilz) [inline crates/lib/src/oracle/jupiter.rs:181]: does this need to be safe division?

File: crates/lib/src/token/token.rs [hunk 6]
```diff
@@ -140,6 +140,119 @@ impl TokenUtil {
...
+            mint_decimals.insert(*mint, decimals);
+        }
+
+        // Calculate total value
+        let mut total_lamports = 0u64;
+
+        for (mint, transfers) in mint_to_transfers.iter() {
+            let price = prices
+                .get(&mint.to_string())
+                .ok_or_else(|| KoraError::RpcError(format!("No price data for mint {mint}")))?;
```
Comment (amilz) [inline crates/lib/src/token/token.rs:259]: This could be annoying, but I think it's the right move...

File: tests/rpc/fee_estimation.rs [hunk 7]
```diff
@@ -345,3 +349,82 @@ async fn test_estimate_fee_comprehensive_with_token_accounts_creation() {
         expected_minimum_fee + 50_000
     );
 }
+
+#[tokio::test]
+async fn test_estimate_fee_with_spl_token_transfer_from_fee_payer() {
```
Comment (amilz) [inline tests/rpc/fee_estimation.rs:354]: I think since our functionality supports multiple...it's probably worth it to add a test for  handling muliple token transfer ix's. i think you could just do this w/ just 2 USDC ixns for testing. or adding a WSOL transfer

File: crates/lib/src/token/token.rs (hunk 3 above)
Comment (dev-jodee) [reply crates/lib/src/token/token.rs:249]: Yeah we can do that in the future, could use redis on disk like we do with the usage limit

File: crates/lib/src/oracle/jupiter.rs (hunk 4 above)
Comment (dev-jodee) [reply crates/lib/src/oracle/jupiter.rs:104]: Crashes, added a test to validate

File: crates/lib/src/oracle/jupiter.rs (hunk 5 above)
Comment (dev-jodee) [reply crates/lib/src/oracle/jupiter.rs:181]: f64 doesn't have checked, but I added some extra checks to make sure with divide by 0 (if somehow it would happen)

File: crates/lib/src/token/token.rs (hunk 6 above)
Comment (dev-jodee) [reply crates/lib/src/token/token.rs:259]: Yeah don't think continuing is safe to not have it sadly

File: crates/lib/src/token/token.rs (hunk 2 above)
Comment (dev-jodee) [reply crates/lib/src/token/token.rs]: I'll add deriving ATA but this won't wrok for non ATA token accounts, think I'll just error out here to fail safe
